pkg/server/plugin/bundlepublisher/awsrolesanywhere/awsrolesanywhere.go

```diff
@@ -0,0 +1,257 @@
+	}
+
+	currentBundle := p.getBundle()
+	if proto.Equal(req.GetBundle(), currentBundle) {
+		// Bundle not changed. No need to publish.
+		return &bundlepublisherv1.PublishBundleResponse{}, nil
+	}
+
+	formatter := bundleformat.NewFormatter(req.GetBundle())
+	bundleBytes, err := formatter.Format(bundleformat.PEM)
+	bundleStr := string(bundleBytes)
+
+	// Check whether there already exists a trust anchor with the name requested
+	// If so, perform an update of its trust bundle
+	var trustAnchor rolesanywheretypes.TrustAnchorDetail
+	foundTrustAnchor := false
+	prevNextToken := ""
+	for ok := true; ok; {
+		// List trust anchors
+		listTrustAnchorsInput := rolesanywhere.ListTrustAnchorsInput{}
+		if prevNextToken != "" {
+			listTrustAnchorsInput.NextToken = &prevNextToken
+		}
+		listTrustAnchorsOutput, err := p.rolesAnywhereClient.ListTrustAnchors(ctx, &listTrustAnchorsInput)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "failed to list trust anchors: %v", err)
+		}
+
+		// Iterate through trust anchors in response
+		for _, curTrustAnchor := range listTrustAnchorsOutput.TrustAnchors {
+			if *curTrustAnchor.Name == config.TrustAnchorName {
+				trustAnchor = curTrustAnchor
+				foundTrustAnchor = true
+				break
+			}
+		}
+
+		if foundTrustAnchor {
+			break
+		}
+
+		if listTrustAnchorsOutput.NextToken == nil {
+			break
+		}
+		prevNextToken = *listTrustAnchorsOutput.NextToken
+	}
+
+	trustAnchorArn := ""
+	if foundTrustAnchor {
+		// Update the trust anchor that was found
+		updateTrustAnchorInput := rolesanywhere.UpdateTrustAnchorInput{
+			TrustAnchorId: trustAnchor.TrustAnchorId,
+			Source: &rolesanywheretypes.Source{
+				SourceType: rolesanywheretypes.TrustAnchorTypeCertificateBundle,
+				SourceData: &rolesanywheretypes.SourceDataMemberX509CertificateData{
+					Value: bundleStr,
+				},
+			},
+		}
+		updateTrustAnchorOutput, err := p.rolesAnywhereClient.UpdateTrustAnchor(ctx, &updateTrustAnchorInput)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "failed to update trust anchor: %v", err)
+		}
+		trustAnchorArn = *updateTrustAnchorOutput.TrustAnchor.TrustAnchorArn
+	} else {
+		// Create a new trust anchor, since an existing one with the requsted name couldn't be found
+		createTrustAnchorInput := rolesanywhere.CreateTrustAnchorInput{
+			Name: &config.TrustAnchorName,
+			Source: &rolesanywheretypes.Source{
+				SourceType: rolesanywheretypes.TrustAnchorTypeCertificateBundle,
+				SourceData: &rolesanywheretypes.SourceDataMemberX509CertificateData{
+					Value: bundleStr,
+				},
+			},
+			Enabled: func() *bool { b := true; return &b }(),
+		}
+
+		createTrustAnchorOutput, err := p.rolesAnywhereClient.CreateTrustAnchor(ctx, &createTrustAnchorInput)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "failed to create trust anchor: %v", err)
+		}
+		trustAnchorArn = *createTrustAnchorOutput.TrustAnchor.TrustAnchorArn
+	}
```

Comment:
Giving the responsibility to create the trust anchor to the plugin has a lot of disadvantages in my opinion.
It makes the publishing logic a lot more complex than what would be needed if it assumes that the trust anchor already exists, but more importantly, it's prone to configuration errors that can lead to create a trust anchor with a wrong name, and not update the proper trust anchor. The fact that there could be multiple trust anchors with the same name doesn't help either, and contributes to make this more error prone.
It also requires to grant additional permissions, when we are only interested in keeping a trust anchor updated. 
I think that's better to give the responsibility to create the trust anchor to the user, and just configure the plugin with a specific trust anchor to update. Only the required permissions to update a trust anchor would be required in this case.

Reply:
Yeah, that makes sense. I will change it so that it works as you suggested in my next revision.